drivers/bus/acpi/compbatt/compbatt.c

```diff
@@ -65,32 +67,351 @@ CompBattSystemControl(IN PDEVICE_OBJECT DeviceObject,
+        BatteryClassStatusNotify(DeviceExtension->ClassData);
+
+        /* Setup the necessary I/O data to query the battery tag */
+        IoStackLocation->Parameters.DeviceIoControl.IoControlCode = IOCTL_BATTERY_QUERY_TAG;
+        IoStackLocation->Parameters.DeviceIoControl.InputBufferLength = sizeof(ULONG);
+        IoStackLocation->Parameters.DeviceIoControl.OutputBufferLength = sizeof(ULONG);
+        BatteryData->Mode = COMPBATT_QUERY_TAG;
+        BatteryData->WorkerBuffer.WorkerTag = 0xFFFFFFFF;
+
+        /* Dispatch our request now to the battery's driver */
+        goto DispatchRequest;
+    }
+
+    /* Our I/O request has been completed successfully, check what did we get */
+    Mode = BatteryData->Mode;
+    switch (Mode)
+    {
+        case COMPBATT_QUERY_TAG:
+        {
+            /*
+             * This battery has just gotten a tag, acknowledge the composite battery
+             * about that so it can recalculate its own composite tag.
+             */
+            if (CompBattDebug & COMPBATT_DEBUG_WARN)
+            {
+                DbgPrint("CompBatt: Battery (Device 0x%p) has a tag of %lu\n", DeviceObject, BatteryData->WorkerBuffer.WorkerTag);
+            }
+
+            /* Ensure the battery tag is not bogus, getting a tag of 0 is illegal */
+            ASSERT(BatteryData->WorkerBuffer.WorkerTag != 0);
+
+            /* Assign the battery tag */
+            BatteryData->Tag = BatteryData->WorkerBuffer.WorkerTag;
+            BatteryData->Flags |= COMPBATT_TAG_ASSIGNED;
+
+            /* Punt the composite battery flags, as the previous cached data no longer applies */
+            DeviceExtension->Flags = 0;
+
+            /* Notify the Battery Class driver this battery has got a tag */
+            BatteryClassStatusNotify(DeviceExtension->ClassData);
+            break;
+        }
+
+        case COMPBATT_READ_STATUS:
+        {
+            /*
+             * Read the battery status only if the IRP has not been cancelled,
+             * otherwise the request must be re-issued again. This typically
+             * happens if the wait values are in conflict which it might
+             * end up in inconsistent battery status results.
+             */
+            if (Status != STATUS_CANCELLED &&
+                !Irp->Cancel)
+            {
+                /*
+                 * If we reach here then the battery has entered into a change of
+                 * power state or its charge capacity has changed.
+                 */
+                if (CompBattDebug & COMPBATT_DEBUG_WARN)
+                {
+                    DbgPrint("CompBatt: Battery state (Device 0x%p) has changed\n", DeviceObject);
+                }
+
+                /* Copy the battery status of this battery */
+                RtlCopyMemory(&BatteryData->BatteryStatus,
+                              &BatteryData->WorkerBuffer.WorkerStatus,
+                              sizeof(BATTERY_STATUS));
```

Comment:
```suggestion
                              sizeof(BatteryData->BatteryStatus));
```
as it seems to be done somewhere down below.